services/systemxhr/js/app.js

```diff
@@ -0,0 +1,101 @@
+(function(window) {
+  'use strict';
+
+  function debug(str) {
+    console.log('SystemXHRService -*-:' + str);
+  }
+
+  // Ok, this kinda sucks because most APIs (and settings is one of them) cannot
+  // be accessed from outside the main thread. So basically everything has to go
+  // down to the SW thread, then back up here for processing, then back down to
+  // be sent to the client. Yay us!
+  var _XMLHttpRequests = {};
+  var _listeners = {};
+
+  var onChangeEvents = [
+      'onabort',
+      'onerror',
+      'onload',
+      'onloadend',
+      'onloadstart',
+      'onprogress',
+      'ontimeout',
+      'onreadystatechange'
+    ];
+
+  var processSWRequest = function(channel, evt) {
+    // We can get:
+    // * methodName
+    // * onpropertychange
+    // * createXMLHttpRequest
+    // * addEventListener
+    // * removeEventListener
+    // * dispatchEvent
+    // All the operations have a requestId, and all the operations over
+    // a XMLHttpRequest also include a xhr id.
+    var remotePortId = evt.data.remotePortId;
+    var request = evt.data.remoteData;
+    var requestOp = request.data;
+
+    function _buildResponseHeadersObject(responseHeaders) {
+      var headers = responseHeaders.split(/\n/);
+      var obj = {};
+      // Last item is useless
+      headers.pop();
+      headers.forEach(header => {
+        var trimeHeader = header.trim();
+        var split = trimeHeader.split(/: /);
+        obj[split[0].trim()] = split[1].trim();
+      });
+
+      return obj;
+    }
+
+    function listenerTemplate(evt) {
+      var clonedEvent = window.ServiceHelper.cloneObject(evt, true);
+      clonedEvent.responseHeaders =
+        _buildResponseHeadersObject(evt.target.getAllResponseHeaders());
+      channel.postMessage({
+        remotePortId: remotePortId,
+        data: {
+          id: request.id,
+          event: clonedEvent
+        }
+      });
+    }
+
+    if (requestOp.operation === 'createXMLHttpRequest') {
```

Comment:
I like it more the way we have implemented this on the more recent mediators (having a table of methods and here just doing something like:

methodTable[requestOp.operation] && methodTable[requestOp.operation](channel, request);